### PR TITLE
Show alert severity and age

### PR DESCRIPTION
**Pull request by ErinLMoore**
Displays actual alerts' severity and age. Adds a function to Alerts to arrange alerts data to make it easier to parse.

Screenshot 2023-05-26 at 2 56 17 PM

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204597186510077

### CODE REVIEW

File: lib/alerts.ex [hunk 1]
```diff
@@ -37,6 +37,30 @@ defmodule Alerts do
...
+          route_id => [
+            %{
+              severity: alert.severity,
+              created_at: alert.created_at,
+              id: alert.id
+            }
+          ]
+        })
+      end)
+      |> Map.merge(acc, fn _k, route_ids_map, acc ->
```
Comment (arkadyan) [inline lib/alerts.ex:58]: thought: So much nested data! It definitely took me a bit to figure out what's going on here (but I also don't know that I have anything better to suggest).

File: lib/alerts_viewer_web/live/bus_live.ex [hunk 2]
```diff
@@ -121,13 +127,32 @@ defmodule AlertsViewerWeb.BusLive do
     """
   end
 
+  def severity_to_minutes(severity) when severity < 3, do: "<10"
+  def severity_to_minutes(severity) when severity >= 9, do: "60+"
+
+  def severity_to_minutes(severity) do
```
Comment (arkadyan) [inline lib/alerts_viewer_web/live/bus_live.ex:133]: thought: What you have here works fine, but another approach might be to list each case as an individual function definition with exact pattern matching, e.g.:

    def severity_to_minutes(3), do: "10"

**Review by arkadyan — COMMENTED** [review]
I added a few little suggestions and one question, but fundamentally this seems good. I think the "nil" suggestion and maybe the "entire alert" question are the only two things that are potentially blocking (and maybe I'm just missing something on that second one).

File: lib/alerts.ex (hunk 1 above)
Comment (ErinLMoore) [reply lib/alerts.ex:58]: I think I can rewrite this one to be a bit clearer, got a little carried away here maybe.

File: lib/alerts.ex [hunk 3]
```diff
@@ -37,6 +37,30 @@ defmodule Alerts do
...
+  Return a map with route ids as string keys and a list of alerts as the value
+  """
+  @spec by_route([Alert.t()]) :: map()
+  def by_route(bus_alerts) do
+    Enum.reduce(bus_alerts, %{}, fn alert, acc ->
+      Enum.map(alert.informed_entity, & &1.route)
+      |> Enum.reduce(%{}, fn route_id, route_ids_acc ->
+        Map.merge(route_ids_acc, %{
+          route_id => [
+            %{
```
[the remark this answers is not in the document]
Comment (ErinLMoore) [reply lib/alerts.ex:50]: I had a (weakly held) reason, but that was before I moved this function into lib/alerts. I think it should give you the whole alert, I'll take this bit out.

File: lib/alerts.ex [hunk 4]
```diff
@@ -37,6 +37,30 @@ defmodule Alerts do
   def by_service(alerts, route_type_string),
     do: Enum.filter(alerts, &Alert.matches_service_type(&1, String.to_integer(route_type_string)))
 
+  @doc """
+  Return a map with route ids as string keys and a list of alerts as the value
+  """
+  @spec by_route([Alert.t()]) :: map()
+  def by_route(bus_alerts) do
+    Enum.reduce(bus_alerts, %{}, fn alert, acc ->
+      Enum.map(alert.informed_entity, & &1.route)
```
[the remark this answers is not in the document]
Comment (ErinLMoore) [reply lib/alerts.ex:46]: I added a filter for nil routes and updated the test for it.

File: lib/alerts_viewer_web/live/bus_live.ex (hunk 2 above)
Comment (ErinLMoore) [reply lib/alerts_viewer_web/live/bus_live.ex:133]: 👍🏻

File: lib/alerts.ex [hunk 5]
```diff
@@ -37,6 +37,29 @@ defmodule Alerts do
...
+  Return a map with route ids as string keys and a list of alerts as the value
+  """
+  @spec by_route([Alert.t()]) :: map()
+  def by_route(alerts) do
+    alerts
+    |> Enum.filter(&Enum.any?(&1.informed_entity, fn ie -> !is_nil(ie.route) end))
+    |> Enum.reduce(%{}, fn alert, acc ->
+      routes = Enum.map(alert.informed_entity, & &1.route)
+
+      alert_by_route_id =
```
Comment (arkadyan) [inline lib/alerts.ex:50]: Ah yeah, breaking it up and naming it this way is smart.

File: test/alerts_test.exs [hunk 6]
```diff
@@ -160,4 +160,60 @@ defmodule AlertsTest do
...
+      alert3 = %Alert{
+        id: 3,
+        severity: 5,
+        created_at: now,
+        informed_entity: [
+          %{route: "30", route_type: 3}
+        ]
+      }
+
+      secret_evil_alert = %Alert{
```
Comment (arkadyan) [inline test/alerts_test.exs:196]: 👍

**Review by arkadyan — APPROVED** [review]
👍

**Pull request merged**